OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -1005,7 +1010,10 @@ void BuildUnit(string category, bool buildRandom)
 				ChooseRandomUnitToBuild(queue) :
 				ChooseUnitToBuild(queue);
 
-			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name))
+			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name
+				&& (!Info.UnitLimits.ContainsKey(unit.Name)
+					|| World.Actors.Where(a => a.Owner == Player).Count(a => a.Info.Name == unit.Name) < Info.UnitLimits[unit.Name])))
```

Comment:
Iterating over all world actors every time the AI decides to build a unit won't be fun.

Also at the very least this should be simplified to
```
if (unit != null
	&& Info.UnitsToBuild.ContainsKey(unit.Name)
	&& (!Info.UnitLimits.ContainsKey(unit.Name)
		|| World.Actors.Count(a => a.Owner == Player && a.Info.Name == unit.Name) < Info.UnitLimits[unit.Name]))

```